test/DebugInfo/linetable-codeview.swift

```diff
@@ -0,0 +1,35 @@
+// RUN: %swiftc_driver %s -target x86_64-unknown-windows-msvc -g -debug-info-format=codeview -emit-ir -o - | %FileCheck %s
+// The division will emit an ``unreachable`` with instructions before and after
+// also being emitted from the division. We want to make sure all instructions
+// from the division have the same debug location.
+
+func myDivide(_ a: Int, _ b: Int) -> Int {
+  // CHECK: define {{.*}} i64 @"$S4main8myDivideyS2i_SitF"(i64, i64)
+  // CHECK-NOT: ret
+  // CHECK: call swiftcc void @"$Ss18_fatalErrorMessage__4file4line5flagss5NeverOs12StaticStringV_A2HSus6UInt32VtF"{{.*}}, !dbg ![[DIV:[0-9]+]]
+  // CHECK-NEXT: unreachable, !dbg ![[DIV]]
+  return a / b      // line 11
+}
+
+func myAdd(_ a: Int, _ b: Int) -> Int {
+  // CHECK: define {{.*}} i64 @"$S4main5myAddyS2i_SitF"(i64, i64)
+  // CHECK-NOT: ret
+  // CHECK: call { i64, i1 } @llvm.sadd.with.overflow.i64({{.*}}), !dbg ![[ADD:[0-9]+]]
+  // CHECK: call void @llvm.trap(), !dbg ![[ARTIFICIAL:[0-9]+]]
```

Comment:
This looks suspicious. Shouldn't the trap have the location of the + operator?

Reply:
The problem is that the ``llvm.trap`` is emitted at the end of the function. If ``llvm.trap`` had the same location as the ``a + b``, then CodeView will have two different sets of instructions for that line. WinDbg cannot decide which instructions to break on.

Reply:
I'm probably misunderstanding something here: Are you saying that CodeView does not support two non-contiguous pc ranges with the same source location? How does it deal with optimized code then?

If the trap does not have the source location of the `+`, how are you planning on attributing the crash caused by the trap to the correct source location?

Reply:
As far as I can tell, that is correct. If I try to break on ``a + b`` in WinDbg, I get a dialogue box saying ``The source line given has multiple addresses``. One address is for the addition and one is for the trap. Although, the breakpoint still works if you select the right address.

Reply:
> As far as I can tell, that is correct. If I try to break on a + b in WinDbg, I get a dialogue box saying The source line given has multiple addresses.

That makes it sounds like WinDbg does actually support multiple address ranges per debug location. I *strongly* recommend to attribute the trap with the correct source location, too, because otherwise developers won't be able to make any sense of the crash log when the trap instruction is hit.

Reply:
That does make sense. However, it would be really annoying for a user to have to select an address every time they want to break on a arithmetic instruction. Would it be reasonable to try to put ``llvm.trap`` with the rest of the instructions from that line instead of the end of the function? Or is this best left alone?

Reply:
Since the trap inevitably involves a branch it may be difficult to guarantee that it ends up in one sequence right after the `+` instruction. I'm not sure why Swift wants to put all the traps at the end of the function. Perhaps to improve the locality of the common code? That's a question for the IRGen folks.
An alternative hack for the windows platform would be to assign an arbitrary column number (perhaps 0?) with the trap instructions to force them to get a unique location.

Reply:
I think the locality answer is right. We expect that code to never actually be executed, so it's best to not have it taking up space in the instruction cache.

Reply:
WinDbg actually ignores column locations which is why I've already set all column locations to zero.

Reply:
What if we have something like ``!DILocation(line: (line of llvm.trap from + in stdlib), scope: B, inlinedAt: !(location of + in source))``? This fixes the trouble with setting a breakpoint on ``+`` in WinDbg and the trap can still be traced back to the ``+``.

Reply:
You could implement this by modifying the behavior of the `transparent` keyword. Basically you would need to change MandatoryInlining to behave like PerformanceInlining (but without messing up any of the diagnostic testscases).

Reply:
That would mess up the source location for the regular addition too, which is a no-go.